kedro/config/omegaconf_config.py

```diff
@@ -0,0 +1,252 @@
+
+        # Destructively merge the two env dirs. The chosen env will override base.
+        common_keys = config.keys() & env_config.keys()
+        if common_keys:
+            sorted_keys = ", ".join(sorted(common_keys))
+            msg = (
+                "Config from path '%s' will override the following "
+                "existing top-level config keys: %s"
+            )
+            _config_logger.info(msg, env_path, sorted_keys)
+
+        config.update(env_config)
+
+        if not config:
+            raise MissingConfigException(
+                f"No files of YAML or JSON format found in {base_path} or {env_path} matching"
+                f" the glob pattern(s): {[*self.config_patterns[key]]}"
+            )
+        return config
+
+    def __repr__(self):  # pragma: no cover
+        return (
+            f"OmegaConfLoader(conf_source={self.conf_source}, env={self.env}, "
+            f"config_patterns={self.config_patterns})"
+        )
+
+
+def load_and_merge_dir_config(conf_path: str, patterns: Iterable[str]):
+    """Recursively load and merge all configuration files in a directory using OmegaConf,
+    which satisfy a given list of glob patterns from a specific path.
+
+    Args:
+        conf_path: Path to configuration directory.
+        patterns: List of glob patterns to match the filenames against.
+
+    Raises:
+        MissingConfigException: If configuration path doesn't exist or isn't valid.
+        ValueError: If two or more configuration files contain the same key(s).
+        ParserError: If config file contains invalid YAML or JSON syntax.
+
+    Returns:
+        Resulting configuration dictionary.
+
+    """
+    if not Path(conf_path).is_dir():
+        raise MissingConfigException(
+            f"Given configuration path either does not exist "
+            f"or is not a valid directory: {conf_path}"
+        )
+
+    paths = [
+        Path(each).resolve()
+        for pattern in patterns
+        for each in iglob(f"{str(conf_path)}/{pattern}", recursive=True)
+    ]
+    deduplicated_paths = set(paths)
+    config_files_filtered = [
+        path for path in deduplicated_paths if _is_valid_config_path(path)
+    ]
+
+    config = {}
+    aggregate_config = []
+    seen_file_to_keys = {}  # type: Dict[Path, AbstractSet[str]]
+
+    for config_filepath in config_files_filtered:
+        try:
+            single_config = OmegaConf.load(config_filepath)
+            config = single_config
+        except (ParserError, ScannerError) as exc:
+            assert exc.problem_mark is not None
+            line = exc.problem_mark.line
+            cursor = exc.problem_mark.column
+            raise ParserError(
+                f"Invalid YAML or JSON file {config_filepath}, unable to read line {line}, "
+                f"position {cursor}."
+            ) from exc
+
+        _check_duplicate_keys(seen_file_to_keys, config_filepath, dict(single_config))
```

Comment:
Do we still want to use this method from `common.py` if we will be reworking it?

Reply:
Good question! The reason I used this method from `common.py` is because we do want this behaviour in the `OmegaConfLoader` and it's exactly the same as for the other config loaders. The other methods in `common.py` that I didn't end up using were very indirect and nested calls, so I refactored them for this config loader. In `0.19.0` I envision we'll get rid of `common.py` so we could move `_check_duplicate_keys` inside `omegaconf_config.py`. 

I could've refactored all methods in `common.py` but that felt like unnecessary work since ideally we wouldn't have this file full of helper methods. 

Do you think this makes sense or would you prefer there's no calls to `common.py` methods at all in this class?

Reply:
This makes sense, thank you. I agree we should move the `_check_duplicate_keys` method in `0.19.0` to `omegaconf_config.py`.